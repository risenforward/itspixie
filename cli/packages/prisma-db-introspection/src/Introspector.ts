import { PostgresConnector } from './connectors/PostgresConnector'
import { SdlPrinter } from './SdlPrinter'
import { ClientConfig } from 'pg'
import { PostgresConnectionDetails } from './types/common'

export class Introspector {
  connectionString: string | ClientConfig
  connector: PostgresConnector
  printer = new SdlPrinter()

  constructor(connectionString: PostgresConnectionDetails) {
    this.connectionString = connectionString
    this.connector = new PostgresConnector(connectionString)
  }

  async listSchemas(): Promise<string[]> {
    return this.connector.listSchemas()
  }

  async introspect(schemaName: string): Promise<{ numTables: number; sdl: string }> {
    const tables = await this.connector.listTables(schemaName)
    const sdl = this.printer.print(tables)

    return { numTables: tables.length, sdl }
  }
}
