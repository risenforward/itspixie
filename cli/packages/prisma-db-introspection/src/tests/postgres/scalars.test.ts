import { Introspector } from '../../Introspector'
import { Client } from 'pg'
import { connectionDetails } from './connectionDetails'

function introspect(): Promise<{ numTables: number; sdl: string; }> {
  return new Introspector(connectionDetails).introspect('DatabaseIntrospector')
}

async function testSchema(sql: string) {
  const client = new Client(connectionDetails)
  await client.connect()
  await client.query('DROP SCHEMA IF EXISTS DatabaseIntrospector cascade;')
  await client.query('CREATE SCHEMA DatabaseIntrospector;')
  await client.query('SET search_path TO DatabaseIntrospector;')
  await client.query(sql)

  expect(await introspect()).toMatchSnapshot()

  await client.end()
}

describe('Introspector', () => {
  test('text columns', async () => {
    await testSchema(`CREATE TABLE "Strings" (
      "a" char(1) DEFAULT NULL UNIQUE,
      "b" varchar(255) DEFAULT NULL,
      "c" text DEFAULT 'abc',
      "d" char(1) NOT NULL,
      "e" varchar(255) NOT NULL,
      "f" text NOT NULL,
      constraint foo unique(b)
      );`)
  })

  test('int columns', async () => {
    await testSchema(`CREATE TABLE "Ints" (
      "a" smallint DEFAULT 3,
      "b" integer DEFAULT NULL,
      "c" bigint DEFAULT NULL,
      "d" smallint NOT NULL,
      "e" integer NOT NULL,
      "f" bigint NOT NULL
      );`)
  })

  test('float columns', async () => {
    await testSchema(`CREATE TABLE "Floats" (
      "a" real DEFAULT 3.4,
      "b" double precision DEFAULT NULL,
      "c" float4 DEFAULT NULL,
      "d" float8 DEFAULT NULL,
      "e" real NOT NULL,
      "f" double precision NOT NULL,
      "g" float4 NOT NULL,
      "h" float8 NOT NULL,
      "i" numeric DEFAULT NULL,
      "j" numeric NOT NULL
      );`)
  })

  test('boolean columns', async () => {
    await testSchema(`CREATE TABLE "Booleans" (
      "a" boolean DEFAULT NULL,
      "b" boolean NOT NULL
      );`)
  })

  test('DateTime columns', async () => {
    await testSchema(`CREATE TABLE "Booleans" (
      "a" timestamp DEFAULT NULL,
      "b" timestamp NOT NULL
      );`)
  })

  test('Json columns', async () => {
    await testSchema(`CREATE TABLE "Booleans" (
      "a" json DEFAULT NULL,
      "b" json NOT NULL
      );`)
  })
})
