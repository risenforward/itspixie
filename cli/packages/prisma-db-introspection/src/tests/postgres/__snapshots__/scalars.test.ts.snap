// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Introspector DateTime columns 1`] = `
"type Booleans @pgTable(name: \\"Booleans\\") {
  a: DateTime
  b: DateTime!
}
"
`;

exports[`Introspector Json columns 1`] = `
"type Booleans @pgTable(name: \\"Booleans\\") {
  a: Json
  b: Json!
}
"
`;

exports[`Introspector boolean columns 1`] = `
"type Booleans @pgTable(name: \\"Booleans\\") {
  a: Boolean
  b: Boolean!
}
"
`;

exports[`Introspector float columns 1`] = `
"type Floats @pgTable(name: \\"Floats\\") {
  a: Float @default(value = 3.4)
  b: Float
  c: Float
  d: Float
  e: Float!
  f: Float!
  g: Float!
  h: Float!
  i: Float
  j: Float!
}
"
`;

exports[`Introspector int columns 1`] = `
"type Ints @pgTable(name: \\"Ints\\") {
  a: Int @default(value = 3)
  b: Int
  c: Int
  d: Int!
  e: Int!
  f: Int!
}
"
`;

exports[`Introspector text columns 1`] = `
"type Strings @pgTable(name: \\"Strings\\") {
  a: String @unique
  b: String @unique
  c: String @default(value = \\"abc\\")
  d: String!
  e: String!
  f: String!
}
"
`;
